crates/blockchain-tree/src/block_buffer.rs

```diff
@@ -0,0 +1,164 @@
+use reth_primitives::{BlockHash, BlockNumHash, BlockNumber, SealedBlockWithSenders};
+use std::collections::{btree_map::Entry, hash_map, BTreeMap, HashMap, HashSet};
+
+/// Contains the Tree of pending blocks that are not executed but buffered
+/// While pipeline is is
+#[derive(Debug)]
+pub struct BlockBuffer {
+    /// Blocks ordered by block number inside the BTreeMap.
+    ///
+    /// Note: BTreeMap is used so that we can remove the finalized old blocks
+    /// from the buffer
+    blocks: BTreeMap<BlockNumber, HashMap<BlockHash, SealedBlockWithSenders>>,
+    /// Needed for removal of the blocks. and to connect the potential unconnected block
+    /// to the connected one.
+    parent_to_child: HashMap<BlockHash, HashSet<BlockNumHash>>,
+    /// Current number of blocks found in buffer.
+    num_of_blocks: usize,
+    /// Max number of blocks.
+    max_blocks: usize,
+}
+
+impl BlockBuffer {
+    /// Create new buffer with max limit of blocks
+    pub fn new(limit: usize) -> Self {
+        Self {
+            blocks: Default::default(),
+            parent_to_child: Default::default(),
+            num_of_blocks: 0,
+            max_blocks: limit,
+        }
+    }
+
+    /// Evict block from the buffer
+    ///
+    /// Iterate over blocks from lowest number to the highest
+    /// and evict first block that does not have childrens.
+    ///
+    /// If there is no block like that evict the first oldest one.
+    fn evict_block(&mut self) {
```

Comment:
not sure about this. we should be able to pass last FCU state here and primarily evict the ones that do not attach (with the same subsequent logic)

Reply:
This fn is used internally only to evict the least desirable block in the buffer, as we have the limit on the amount of blocks we are buffering.


There are one function that react on FCU finalization:
`clean_old_blocks` removes any old blocks that are finalized (and their children).
and second reacts on newly received payload:
`take_all_childrens` this function takes all child if there is some of the payload that got included inside tree.

Reply:
What if the least desirable block attaches to the FCU's head hash? All I'm saying is that FCU state should be considered even when internally evicting the blocks

Reply:
I thought you just misunderstood, sorry about that. With the current algorithm, I try to minimize this, considering FCU would be better but would complicate either eviction or buffer structure and the current algo should work.

The assumption is that buffer size should be  bigger than finalization windows, let take 100 slots.

Algorithm is
```
from oldest block to newest block {
     check if block has child
     if it does not, remove it and return
     (i have bug here where i remove newest block)
}
remove oldest block from buffer.
```

Oldest/newest blocks are considering their block number.

So we are prioritising the uncles and shorter "chains", and in the end we would just remove oldest blocks from buffer.

Reply:
If this does not work we could even do a simpler version with LRU to evict the oldest inserted block.

Reply:
I defaulted to LRU it is simpler and more straightforward, and new blocks always come with time delay so we should be okay with it.